### ISSUE TITLE
onBlur functionality in latest version of redux-form not working

### ISSUE THREAD

**Issue by reznord** [erikras/redux-form#2334]
After updating to the latest version of redux-form (v 6.4.1), the onBlur is not working as expected. For example: 

I have the following two Dropdown components where the 2nd dropdown is values are dynamically allocated based on the first dropdown values.

```
constructor(props) {
  super(props);
  this.setDropdownValue = this.setDropdownValue.bind(this);
}

setDropdownValue(option) {
    const routingOptions = this.props[entitiesRoutingMap[option]] || [
      {
        id: null,
        name: 'None',
        disabled: true,
      },
    ];
    this.setState({ identityType: option, routingOptions });
  }

<Field component={Dropdown} valueKey="value" popup options={entities} className="row-popup" name="identityType" labelKey="label" label="Routing Type" description="Select the routing type you need" onBlur={this.setDropdownValue} />
        
<Field component={Dropdown} labelKey="name" valueKey="id" name="identity" popup searchable options={this.state.routingOptions} label="Routing Entry" className="row-popup" description="Some description" />
```

This code works perfectly fine for me with redux-form v6.3.2 . But in the latest version of the redux-form for some reason, this fails.

And when I install the latest version of redux-form, it asks to downgrade the react version and the redux version as well.

Not sure if anyone else is facing the same issue or not.

**Comment by kolpav** [comment]
I have same problem

**Comment by reznord** [comment]
@kolpav I had to revert back to the v6.3.2 for now. It works fine with that version.

**Comment by erikras** [comment]
That functionality was never officially supported. Something like this could be added, but it was never part of the API contract.

**Comment by kolpav** [comment]
@erikras 
I am not sure if I understand correctly `onBlur` is mentioned in docs here

> The Field component will provide your input with onChange, onBlur, onFocus, onDrag, and onDrop props to listen to..

and here

>These are props that Field will pass to your wrapped component
...
input.onBlur(eventOrValue) : Function
A function to call when the form field loses focus. It expects to either receive the React SyntheticEvent or the current value of the field

I assume `onBlur` is part of API. I can't test it right now but I am sure even on the simplest example I could come up with `onBlur` was not called. When I get home I'll try to provide more info.

**Comment by dfbaskin** [comment]
This seems to be the change that breaks things:

From v6.3.2 (`render` function in `CreateField.js`)

```javascript
      const { custom, ...props } = createFieldProps(getIn,
        name,
        {
          ...rest,
          name,
          blur,
          change,
          focus
        },
        asyncValidate
      )
```

v6.4.1

```javascript
      const { custom, ...props } = createFieldProps({ getIn, toJS },
        name,
        {
          ...rest,
          onBlur: this.handleBlur,
          onChange: this.handleChange,
          onDrop: this.handleDrop,
          onDragStart: this.handleDragStart,
          onFocus: this.handleFocus
        }
      )
```

So the `onBlur`, `onFocus`, etc. events are now intercepted by the `Field` component and translated into actions (`_reduxForm.blur()`, `_reduxForm.focus()`, etc.) Previously you could pass handlers down to the control itself through the `Field` component.

**Comment by kolpav** [comment]
@erikras 
I think I understand now what you meant by that it was never officially supported. `Field` will provide `onBlur`, `onFocus` etc. to your component only for dispatching correct redux-form action. So now question is how do you execute logic on i.e `onBlur` event? With `reducer.plugin` somehow?

**Comment by semy** [comment]
Same problem, we can not use version 6.4.2 because none of our forms will work... Maybe we could nevertheless find a solution? :)

**Comment by semy** [comment]
Okay i fixed the issue by myself...

For all who have the same problem, here is my fix:
```
class FieldComponent extends Component {
  render() {
    const { styles, component, ref, ...props } = this.props;
    const { onFocus, onBlur, onChange, onDragStart, onDrop } = this.props;

    const meta = {
      onFocus,
      onBlur,
      onChange,
      onDragStart,
      onDrop,
    };

    return (
      <div styleName="text">
        <Field component={!component ? renderField : component} meta={meta} {...props} />
      </div>
    );
  }
}
```

Then in my `renderField`:

```
class renderField extends Component {
  constructor(props) {
    super(props);
    this.state = { autoFocused: null };
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onDragStart = this.onDragStart.bind(this);
    this.onDrop = this.onDrop.bind(this);
  }

...

  onFocus(e) {
    const { input, meta: { onFocus } } = this.props;

    if (onFocus && typeof onFocus === 'function') {
      onFocus(e, input);
    }
  }

  onBlur(e) {
    const { input, meta: { onBlur } } = this.props;

    if (onBlur && typeof onBlur === 'function') {
      onBlur(e, input);
    }
  }

  /* I use material ui and the "k" is for the key of a select onChange */
  onChange(e, v, k) {
    const { input, meta: { onChange } } = this.props;

    if (onChange && typeof onChange === 'function') {
      onChange(e, v, input, k);
    }
  }

  onDragStart(e) {
    const { input, meta: { onDragStart } } = this.props;

    if (onDragStart && typeof onDragStart === 'function') {
      onDragStart(e, input);
    }
  }

  onDrop(e) {
    const { input, meta: { onDrop } } = this.props;

    if (onDrop && typeof onDrop === 'function') {
      onDrop(e, input);
    }
  }

...

};
```

And then set the following props for any child field components (in this case material ui):
```
returnField = <TextField
          ...
          onFocus={(e) => { inputOnFocus(e); this.onFocus(e); }}
          onBlur={(e) => { inputOnBlur(e); this.onBlur(e); }}
          onChange={(e, v) => { inputOnChange(v); this.onChange(e, v); }}
          onDragStart={(e) => { inputOnDragStart(e); this.onDragStart(e); }}
          onDrop={(e) => { inputOnDrop(e); this.onDrop(e); }}
          {...input}
          {...props}
        />;
```

Select field of material ui differs a little bit:

```
returnField = <SelectField
        ...
        onFocus={(e) => { inputOnFocus(e); this.onFocus(e); }}
        onBlur={(e) => { inputOnBlur(e); this.onBlur(e); }}
        onChange={(e, k, v) => { inputOnChange(v); this.onChange(e, v, k); }}
        onDragStart={(e) => { inputOnDragStart(e); this.onDragStart(e); }}
        onDrop={(e) => { inputOnDrop(e); this.onDrop(e); }}
        {...input}
        {...props}
      >
        {children}
      </SelectField>;
```

Hope this will help :)

**Comment by semy** [comment]
Maybe you use `data-meta` instead of `meta` :-)

And then on destructuring `const { input, 'data-meta': { onBlur } } = this.props;`

Was there not a way to set own meta-values?

**Comment by erikras** [comment]
You can pass any prop that doesn't collide with `input`, `name`, or `meta`, and it will get passed down to your component.

**Comment by semy** [comment]
Okay perfect, thanks for your help :-)

**Comment by kolpav** [comment]
@erikras Would it be possible to add that unintentional functionality back please or would you accept PR? It was all nice and clean before but now we need workarounds like @semy showed us and it just doesn't feel right IMHO.

**Comment by dmason30** [comment]
IMO I think fields need to be able to support custom callbacks for 

onFocus,
      onBlur,
      onChange,
      onDragStart,
      onDrop,

In my instance I have two numeric fields that onChange using certain logic changes the values of another field and then does some rounding onBlur of that field. When I switch to the latest they no longer change each other as only the action creaters are being called. 

Inside my field compontent I was doing this to make it work:

                            input.onBlur = (e) => {
                                input_copy.onBlur(e);
                                properties.onBlur(e);
                            };

But clearly this no longer works.

**Comment by frankjwu** [comment]
[Field documentation](http://redux-form.com/6.4.2/docs/api/Field.md/) contains the following:

> 1. The name prop is required. It is a string path, in dot-and-bracket notation, corresponding to a value in the form values. It may be as simple as 'firstName' or as complicated as contact.billing.address[2].phones[1].areaCode.
> 2. The component prop is required. It may be a Component, a stateless function, or string name of one of the default supported DOM inputs (input, textarea or select), See Usage section below. To learn about the props that will provided to whatever your component is, see the Props section below.
> 3. All other props will be passed along to the element generated by the component prop.

This seems to suggest to me that onBlur, onChange, etc. will be passed downwards to the component without the need for a separate wrapper like @semy's.

What is the recommended approach moving forward?

**Comment by erikras** [comment]
> All other props will be passed along to the element generated by the component prop.

@frankjwu Right, except that these specific five props are the way that Redux Form hooks into the input, so they are not included in that "All other props will be passed along" line. It needs updating.

**Comment by srinadh239** [comment]
I know its been updated in the master, but I am waiting for the new version release.

For the time being, I had done a work around. Hope it helps.

In the input field onBlur, call the custom function as well including input.onBlur().

```javascript
onBlur={(...args) => {input.onBlur(...args); customFunction(...args)}}
```

**Comment by erikras** [comment]
Fix released as [`v6.5.0`](https://github.com/erikras/redux-form/releases/tag/v6.5.0).

**Comment by reznord** [comment]
@erikras thanks a lot for the fix. Works awesome :)
